src/Common/Visuals/Avalonia/MorganStanley.ComposeUI.Common.VisualUtils/DispatcherAdapter.cs

```diff
@@ -0,0 +1,51 @@
+﻿/// ********************************************************************************************************
+///
+/// Morgan Stanley makes this available to you under the Apache License, Version 2.0 (the "License").
+/// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+/// See the NOTICE file distributed with this work for additional information regarding copyright ownership.
+/// Unless required by applicable law or agreed to in writing, software distributed under the License
+/// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and limitations under the License.
+/// 
+/// ********************************************************************************************************
+
+using Avalonia.Threading;
+using MorganStanley.ComposeUI.Common.Interfaces;
+using NP.Utilities.Attributes;
+
+
+namespace MorganStanley.ComposeUI.Common.VisualUtils
+{
+    /// <summary>
+    /// this class implements Avalonia UI thread marshalling functionality. 
+    /// </summary>
+    [Implements(typeof(ISyncContext), isSingleton:true)]
+    public class DispatcherAdapter : ISyncContext
+    {
+        /// <summary>
+        /// Returns true, if the current thread the same as the Avalonia UI thread (no marshaling is needed),
+        /// false otherwise.
+        /// </summary>
+        public bool CheckAccess()
+        {
+            return Dispatcher.UIThread.CheckAccess();
+        }
+
+        /// <summary>
+        /// performs the action within the Avalonia UI thread
+        /// </summary>
+        /// <param name="action">Action to perform</param>
+        /// <param name="priority">Thread priority</param>
+        public void Post(Action action, SyncPriority priority = SyncPriority.Normal)
+        {
+            if (!CheckAccess())
```

Comment:
I see that this is an attempt at efficiency.  However, it is also a violation of the Law of Least Surprise.  If I call `_syncContext.Post(MyWork, SyncPriority.Input)`, and I know there is queued work at higher priority that should run first, then I expect MyWork will not run until after it.  The reality is that if we are already on the UI thread the SyncPriority is ignored effectively bypassing the priority queueing.

If performance/responsiveness of applications reveals the need for something like this, then I prefer we define a separate method on the interface with a name and documentation that conveys this behaviour. Or if the interface has CheckAccess then this can be an extension method and doesn't have to be on the interface.

A real world example of how this current behaviour can bite.  Using a queued operation at Input priority we can check the difference in time between just prior to the enqueue and when the callback happens at Input priority as a calculated estimate of the UI responsiveness of an application.  If we have a service that does this periodically to monitor the Dispatcher performance, and it runs on the UI thread, and we inject this implementation, then it will appear as though our applications are performing perfectly well even if they are not.

Reply:
I think your suggestion is a bit of an overkill for such simple case. This is how such methods are usually resolved with Dispatcher or SyncContext. I think we do not need to make it overly complex in order to increase consistency.